File: Java/p10305.java
import java.util.Scanner;
import java.util.Stack;
import java.util.LinkedList;
import java.util.Iterator;

public class p10305 {
	public static void main (String [] args) {
		Scanner input = new Scanner(System.in);
		while (true) {
			int no = input.nextInt();
			int command = input.nextInt();
			
			if (no == 0 && command == 0) break;
			
			Graph g = new Graph(no);
			
			for (int i = 0; i < command; i++) {
				int awal = input.nextInt();
				int akhir = input.nextInt();
				g.addEdge(awal-1, akhir-1);
			}
			
			g.topologicalSort();
			System.out.println();
		}
	}
	
	static class Graph
	{
		private int V;
		private LinkedList<Integer> adj[];
	 
		Graph(int v)
		{
			V = v;
			adj = new LinkedList[v];
			for (int i=0; i<v; ++i)
				adj[i] = new LinkedList();
		}
	 
		void addEdge(int v,int w) { adj[v].add(w); }
	 
		void topologicalSortUtil(int v, boolean visited[],
								 Stack stack)
		{
			visited[v] = true;
			Integer i;
	 
			Iterator<Integer> it = adj[v].iterator();
			while (it.hasNext())
			{
				i = it.next();
				if (!visited[i])
					topologicalSortUtil(i, visited, stack);
			}
	 
			stack.push(new Integer(v));
		}
	 
		void topologicalSort()
		{
			Stack stack = new Stack();
	 
			boolean visited[] = new boolean[V];
			for (int i = 0; i < V; i++)
				visited[i] = false;
	 
			for (int i = 0; i < V; i++)
				if (visited[i] == false)
					topologicalSortUtil(i, visited, stack);
	 
	        boolean temp = true;
	        
			while (stack.empty()==false) {
			    if (temp) {
			        System.out.print(((int)stack.pop())+1);
			        temp = false;
			    } else {
			        System.out.print(" " + (((int)stack.pop())+1));
			    }
			}
		}
	}
}

File: C,C++/p10305.cpp
#include<iostream>
#include <list>
#include <stack>
using namespace std;
 
class Graph
{
    int V;
 
    list<int> *adj;
 
    void topologicalSortUtil(int v, bool visited[], stack<int> &Stack);
public:
    Graph(int V);
 
    void addEdge(int v, int w);
 
    void topologicalSort();
};
 
Graph::Graph(int V)
{
    this->V = V;
    adj = new list<int>[V];
}
 
void Graph::addEdge(int v, int w)
{
    adj[v].push_back(w);
}
 
void Graph::topologicalSortUtil(int v, bool visited[], 
                                stack<int> &Stack)
{
    visited[v] = true;
 
    list<int>::iterator i;
    for (i = adj[v].begin(); i != adj[v].end(); ++i)
        if (!visited[*i])
            topologicalSortUtil(*i, visited, Stack);
 
    Stack.push(v);
}
 
void Graph::topologicalSort()
{
    stack<int> Stack;
 
    bool *visited = new bool[V];
    for (int i = 0; i < V; i++)
        visited[i] = false;
 
    for (int i = 0; i < V; i++)
      if (visited[i] == false)
        topologicalSortUtil(i, visited, Stack);
 
    bool temp = true;
 
    while (Stack.empty() == false)
    {
        if (temp) {
            cout << Stack.top()+1;
            Stack.pop();
            temp = false;
        } else {
            cout << " " << Stack.top()+1;
            Stack.pop();
        }
    }
}
 
int main()
{
    while (true) {
        int no, command;
        cin >> no >> command;
        if (no == 0 && command == 0) break;
        
        Graph g(no);
        while (command--) {
            int mulai, tujuan;
            cin >> mulai >> tujuan;
            g.addEdge(mulai-1, tujuan-1);
        }
        
        g.topologicalSort();
        cout << endl;
    }
    return 0;
}